---
layout: page
title: ELB18S
tagline: Entry Level Bioinformatics (Second Course in 2018)
description: Entry level course with a soft introduction to NGS data analysis 
---

## Course Description
This is an **entry level** course aimed that those with a reasonable biological background but **no significant experience with bioinformatics**. The course is broadly based around a series of exercises in which a combination of simple analytical tools and reference to publicly available databases is applied to the investigation of a single human gene. The training manual for the course is comprised of detailed instructions for the tasks undertaken. Included are, questions (with answers) and discussion of and the interpretation of the results achieved.

Participants are asked to imagine an interest in the **disease aniridia**. Course exercises then provide extremely detailed instruction leading participants to discover the gene primarily associated with this disease and all that is interesting about that gene and its protein products.

This course will also provide a **soft introduction to Next Generation Sequencing (NGS) data analysis**. This part of the course aims at providing basic skills that are needed when one needs to process NGS data, such as evaluating data quality, trimming sequences, changing data formats, visualising data, etc. Then, participants will learn how to address a simple transcriptomics problem, stepwise, using open source bioinformatics tools.

## Target Audience
This course is intended for those wishing to investigate how they might begin to exploit the ever expanding abundance of computing and data resources for researchers seeking help in using them. 

## Detailed Program

### Day 1
[**1**: Small introduction of what is Bioinformatics](assets/000-Bioinformatics_Definition.pdf)

[**2**: Genome Databases and Tools](assets/01-Databases_Practical.pdf)

### Day 2
[**3**: Graphical and Textual Pairwise ALignments](assets/02-Pairwise_Alignment_Practical.pdf)

[**4**: Databases Searching Methods (primarly blast)](assets/03-Database_Searching_Practical.pdf)

### Day 3
[**5**: Primer Design](assets/04-Primer_Design_Practical.pdf)

[**6**: Simple Protein Sequence Analysis, Secondary Structure Prediction and Protein Domain/Motifs Databases](assets/05-Structure_Prediction_Practical.pdf)

[**7**: Multiple Sequence Alignment](assets/06-Multiple_Sequence_Alignment_Practical.pdf)

### Day 4
[**8**: Broadly describe the High Throughput Sequencing Workflow](pages/L08.md)

[**9**: Interpret and Manipulate raw sequencing data](pages/L09.md)
  + [**9.1**: The FastQ file format](pages/L09.md/#LO9.1)
  + [**9.2**: Use FastQC to analyse the quality of data in a fastq file](pages/L09.md/#LO9.2)
  + [**9.3**: Use Trimmomatic to improve the quality of data in a fastq file](pages/L09.md/#LO9.3)

[**10**: Align HTS data against a genome](pages/L10.md)
  + [**10.1**: Use the BWA aligner to align HTS data against a genome](pages/L10.md/#L10.1)
  + [**10.2**: The SAM/BAM alignment format](pages/L10.md/#L10.2)

### Day 5
[**11**: Visualize alignments](pages/L11.md)
  + [**11.1**: Use Qualimap to assess quality of alignments](pages/L11.md/#L11.1)
  + [**11.2**: Use IGV to visualize the content of a BAM file](pages/L11.md/#L11.2)

[**12**: Broadly describe different HTS applications](pages/L12.md)
  + [**12.1**: Variant detection in resequencing experiments](pages/L12.md/#L12.1)
  + [**12.2**: Denovo genome assembly and annotation](pages/L12.md/#L12.2)
  + [**12.3**: Transcriptomics using RNA-Seq](pages/L12.md/#L12.3)
  + [**12.4**: 16S Metagenomics](pages/L12.md/#L12.4)
  + [**12.5**: Epigenetics](pages/L12.md/#L12.5)

---

### [Learning objectives](pages/learning_objectives.md)


### [Instructors](pages/instructors.md)
---

The source for this course webpage is [on github](https://github.com/GTPB/Web_course_template).
